test/bio_enc_test.c

```diff
@@ -12,127 +12,199 @@
 #include <openssl/bio.h>
 #include <openssl/rand.h>
 
-int main()
+#include "test_main.h"
+#include "testutil.h"
+
+#define ENCRYPT  1
+#define DECRYPT  0
+
+#define DATA_SIZE    1024
+#define MAX_IV       32
+#define BUF_SIZE     (DATA_SIZE + MAX_IV)
+
+static const unsigned char KEY[] = {
+    0x51, 0x50, 0xd1, 0x77, 0x2f, 0x50, 0x83, 0x4a,
+    0x50, 0x3e, 0x06, 0x9a, 0x97, 0x3f, 0xbd, 0x7c,
+    0xe6, 0x1c, 0x43, 0x2b, 0x72, 0x0b, 0x19, 0xd1,
+    0x8e, 0xc8, 0xd8, 0x4b, 0xdc, 0x63, 0x15, 0x1b
+};
+
+static const unsigned char IV[] = {
+    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08
+};
+
+static int do_bio_cipher(const EVP_CIPHER* cipher, const unsigned char* key,
+    const unsigned char* iv)
 {
     BIO *b;
-    static const unsigned char key[16] = { 0 };
-    static unsigned char inp[1024] = { 0 };
-    unsigned char out[1024], ref[1024];
+    static unsigned char inp[BUF_SIZE] = { 0 };
+    unsigned char out[BUF_SIZE], ref[BUF_SIZE];
     int i, lref, len;
 
     /* Fill buffer with non-zero data so that over steps can be detected */
-    if (RAND_bytes(inp, sizeof(inp)) <= 0)
-        return -1;
+    if (!TEST_int_gt(RAND_bytes(inp, DATA_SIZE), 0))
+        return 0;
 
-    /*
-     * Exercise CBC cipher
-     */
+    /* Encrypt tests */
 
     /* reference output for single-chunk operation */
     b = BIO_new(BIO_f_cipher());
-    if (!BIO_set_cipher(b, EVP_aes_128_cbc(), key, NULL, 0))
-        return -1;
-    BIO_push(b, BIO_new_mem_buf(inp, sizeof(inp)));
+    if (!TEST_int_eq(BIO_set_cipher(b, cipher, key, iv, ENCRYPT), 1))
+        return 0;
+    BIO_push(b, BIO_new_mem_buf(inp, DATA_SIZE));
     lref = BIO_read(b, ref, sizeof(ref));
     BIO_free_all(b);
 
     /* perform split operations and compare to reference */
     for (i = 1; i < lref; i++) {
         b = BIO_new(BIO_f_cipher());
-        if (!BIO_set_cipher(b, EVP_aes_128_cbc(), key, NULL, 0))
-             return -1;
-        BIO_push(b, BIO_new_mem_buf(inp, sizeof(inp)));
+        if (!TEST_int_eq(BIO_set_cipher(b, cipher, key, iv, ENCRYPT), 1))
+             return 0;
+        BIO_push(b, BIO_new_mem_buf(inp, DATA_SIZE));
         memset(out, 0, sizeof(out));
         out[i] = ~ref[i];
         len = BIO_read(b, out, i);
         /* check for overstep */
-        if (out[i] != (unsigned char)~ref[i]) {
-            fprintf(stderr, "CBC output overstep@%d\n", i);
-            return 1;
-        }
+        if (!TEST_uchar_eq(out[i], (unsigned char)~ref[i]))
```

Comment:
Probably a "TEST_info()" call might be appropriate to tell us the value of i in the event of a failure.